### ISSUE TITLE
index.yaml not regenerated when chart added to s3 bucket

### ISSUE THREAD

**Issue by PeteMac88** [helm/chartmuseum#444]
Hey there,
in the documentation under [Notes on index.yaml](https://github.com/helm/chartmuseum#notes-on-indexyaml) it says that chartmuseum should recognize if a chart is added/delete to the s3 bucket. In our use case the CD pipelines push helm charts directly to the s3 bucket and we only use chartmuseum to access these helm charts in a k8s cluster. Our problem is that the new charts are not recognized by chartmuseum and the index.yaml updates only if the pod is restarted. 

Here are our values for the chartmuseum helm chart
```
env:
      open:
        STORAGE: amazon
        STORAGE_AMAZON_BUCKET: "*****"
        STORAGE_AMAZON_PREFIX: "******"
        STORAGE_AMAZON_REGION: "eu-central-1"
        DEBUG: true
      existingSecret: chartmuseum-s3
      existingSecretMappings:
        AWS_ACCESS_KEY_ID: ******
        AWS_SECRET_ACCESS_KEY:  ******
      securityContext:
        enabled: true
        fsGroup: 65534
```

Is there any magic trick to enable this because it would be a deal breaker for us

**Comment by m1pl** [comment]
We solved this for now by adding `--cache-interval=1m` to chartmuseum. This shouldn't be necessary according to the docs:
`[...] upon each request, the storage backend is scanned for changes compared to the cache.`

Any tips as to why it's not happening in our case are welcome.

**Comment by jdolitsky** [comment]
@PeteMac88 - can you share some information about the chart uploaded? (version number etc.)

Also, is there anything about your s3 bucket that is different (using minio etc)

**Comment by m1pl** [comment]
@jdolitsky It's a plain s3 bucket, no minio. The problem is with every chart, even a fresh one generated with `helm create`. Could you please point us to the code where the _index.yaml_ is supposed to be regenerated upon every request? I couldn't find it, maybe I'm missing something?

**Comment by m1pl** [comment]
@jdolitsky any updates? We can further debug this issue if you could point us in the right direction. As far as I can tell from the code, index.yaml only gets regenerated in multitenancy mode.

**Comment by jdolitsky** [comment]
@m1pl - could you try deleting `index-cache.yaml` from your storage bucket? People have had issues with this getting out of sync in the past.

**Comment by m1pl** [comment]
@jdolitsky already tried, doesn't help.

**Comment by jdolitsky** [comment]
Can you enable debug logging and share the log contents after uploading a chart?

**Comment by zware** [comment]
Our project is just starting to try to replace `helm serve` with `chartmuseum`, and we're seeing the same issue with local storage; `index.yaml` does not seem to be updated when changes are made to the repo directory.  Debug logging doesn't seem to show any attempt to refresh the index when requesting `index.yaml` (redacted and de-duplicated):

```
$ chartmuseum --version
ChartMuseum version v0.13.1 (build 79bb39c)
$ chartmuseum --port 8879 --storage local --storage-local-rootdir ~/.helm/repository/local/ --context-path /charts --debug
2021-04-12T11:35:51.704-0500    DEBUG   Fetching chart list from storage        {"repo": ""}
2021-04-12T11:35:51.704-0500    DEBUG   GetObject From Storage Backend  {"repo": "", "path": "redacted-1.1.0.tgz"}
2021-04-12T11:35:51.704-0500    DEBUG   Regenerating index.yaml {"repo": ""}
2021-04-12T11:35:51.704-0500    DEBUG   Loading charts packages from storage (this could take awhile)   {"repo": "", "total": 27}
2021-04-12T11:35:51.706-0500    DEBUG   Adding chart to index   {"repo": "", "name": "redacted", "version": "1.1.0"}
2021-04-12T11:35:51.713-0500    DEBUG   index.yaml regenerated  {"repo": ""}
2021-04-12T11:35:51.713-0500    DEBUG   Entry saved in cache store      {"repo": ""}
2021-04-12T11:35:51.713-0500    INFO    Starting ChartMuseum    {"host": "0.0.0.0", "port": 8879}
2021-04-12T11:35:51.713-0500    DEBUG   Starting internal event listener
2021-04-12T11:35:51.713-0500    DEBUG   index-cache.yaml saved in storage       {"repo": ""}
# chart manually copied into ~/.helm/repository/local
# curl http://localhost:8879/charts/index.yaml from separate shell
2021-04-12T11:36:09.118-0500    DEBUG   [1] Incoming request: /charts/index.yaml        {"reqID": "adf...252"}
2021-04-12T11:36:09.118-0500    DEBUG   [1] Entry found in cache store  {"repo": "", "reqID": "adf...252"}
2021-04-12T11:36:09.119-0500    INFO    [1] Request served      {"path": "/charts/index.yaml", "comment": "", "clientIP": "::1", "method": "GET", "statusCode": 200, "latency": "124.746µs", "reqID": "adf...252"}
```

**Comment by williambrode** [comment]
I'm having this same issue with a GCP bucket.  I assumed maybe chart museum needed the edits to come through its API but then I found this in the docs:
```If you manually add/remove a .tgz package from storage, it will be immediately reflected in GET /index.yaml.```

**Comment by peppe77** [comment]
we have just upgraded to v0.13.1 and are also experiencing same issue. any suggestions?

**Comment by williambrode** [comment]
@peppe77 what version were you on before?  If it wasn't happening on that version it may help narrow down what caused the issue.

**Comment by peppe77** [comment]
@williambrode we were on v0.8.3 though due to performance issue (panic) when handling helm repo with lot of charts, we just move up to v0.13.1 (as fix for this performance issue is on v0.13.0). Any work around?

**Comment by williambrode** [comment]
See up above: https://github.com/helm/chartmuseum/issues/444#issuecomment-811091757
This worked for me - though I'm not sure what other negative implications there are (pulling more data from storage?).

**Comment by jdolitsky** [comment]
My apologies that people are experiencing problems trying to perform what should otherwise be a simple upgrade.

I did some testing with Amazon S3. I believe I may have a theory what is happening here-

Starting in v0.13.0, we started using an event emitter pattern. This improves performance by **only updating the index when an upload occurs via the ChartMuseum API**.

If a chart is uploaded using any other means, the index will not be updated. We would accept a PR to add an option such as `--always-update` to always check the backend storage upon `GET /index.yaml` (basically the old pattern).

---

My question to people experiencing this problem:

Instead of using the ChartMuseum upload API, are you all uploading charts using cloud-specific tooling such as `aws s3 cp` or `gsutil cp`, via cloud provider APIs/SDKs, or by performing manual uploads in your cloud's browser UI? If so, this is the root of the problem (not that it isn't a perfectly valid workflow).

**Comment by peppe77** [comment]
@jdolitsky our pipelines are all automated and have been using ChartMuseum api to push helm charts to our multi-tenant setup. Previously, we were on v0.8.3 though heavily impacted by random chartmuseum panics when requests were coming to helm repos with big number of charts. Some research pointed out to a known bug that got fixed in v0.13.0, so we upgraded to v0.13.1 though started to noticed a new random problem, that is, we have seen charts successfully uploaded (via api) though not listed in the index.yaml. thoughts? suggestions?

**Comment by peppe77** [comment]
@m1pl can you pls provide more info regarding the change in v0.13.0? we were left with the impression that index is regenerated upon every request....

**Comment by m1pl** [comment]
@peppe77 I think you got your tags mixed up, I'm not the right person to ask ;) Check out [this](https://github.com/helm/chartmuseum/issues/444#issuecomment-827103530) comment. Apparently the index only gets updated when uploading charts through the ChartMuseum API.

**Comment by m1pl** [comment]
I also posted a solution [here](https://github.com/helm/chartmuseum/issues/444#issuecomment-811091757). This polls every minute and works perfectly fine for us. It should also keep the load acceptable on repositories with lots of requests.

**Comment by peppe77** [comment]
@m1pl sorry for that. we only use ChartMuseum API (v0.13.1) - so , slightly different from your case where you by-pass API (though, agreed documentation states that index should be regenerated). Before v0.13.1, we did not have diff between charts uploaded and index though now on v0.13.1 - we have seen a few cases :(

**Comment by williambrode** [comment]
@jdolitsky is it worth backing this out or marking this release as a bad one until we can get this fixed?  I almost gave up on chart museum entirely because it wasn't working.

**Comment by jdolitsky** [comment]
@williambrode - you might try simply running v0.12.0 and see if the issue goes away until we have a fix in place for this. Is the cache interval causing any issues for your team's workflow?

**Comment by williambrode** [comment]
@jdolitsky from the debug logs it looked like it was redownloading every chart at every interval.  Ideally I'd just like it to use the index.yaml file in storage and not cache anything - but I don't think it's ever supported that right?

**Comment by phntom** [comment]
Solved for S3+KMS users: I had a couple of issues with amazon's s3 and finally figured out chartmuseum's iam role did not have permissions to the KMS used to upload the files (default encryption KMS specified in the bucket).
once added the permission and restarted the processes - index.yaml was properly populated.
Side note, the latest version creates a different s3 path then older versions (I didn't upgrade for a year, I know, shame on me).
previously files under s3://bucketname/ should now be under s3://bucketname/bucketname/ including index-cache.yaml

**Comment by hawksight** [comment]
We had the same issue on GCS as the backend and using `helm-gcs` plugin to push released from the CI, a combination of:

```yaml
env:
  open:
    DISABLE_STATEFILES: true
extraArgs:
  - --cache-interval=1m
```

In the helm chart install does seem to have resolved our issue for now.
Thanks @m1pl for the tip. I would never have found that `--cache-interval` setting without your prompt in https://github.com/helm/chartmuseum/issues/444#issuecomment-811091757

**Comment by stromnet** [comment]
This hit me too. As a new user of chartmuseum, in our case we build externally and push to S3, and have several clusters see these new artifacts. Thus, pushing to (multiple) chartmuseum instances directly would not be optimal.

A potential alternative to scan S3 every $INTERVAL, or scan every time time index is fetched, would be if there is some API endpoint for triggering immediate regeneration of index, in addition to $INTERVAL based refresh. Then I could have let's say 24h interval, but have some mechanism to trigger rescan when new chart is pushed to S3 (for example, via SNS/SQS, but that is probably outside the scope of chartmuseum).

For moment I'll have to live with $INTERVAL and some delays .

**Comment by artem-kosenko** [comment]
works only with `--cache-interval=1m` argument.

Strange, but I did not find this key in official documenation https://chartmuseum.com/docs/#other-cli-options
But it is mentioned on github https://github.com/helm/chartmuseum#cache

**Comment by pradheba-rangarajan** [comment]
we upgraded chartmuseum to v0.14.0 yet the issue still exists. we updated to use `--cache-interval=1m` as suggested above and monitoring if it works

**Comment by williambrode** [comment]
Any way to bump the priority of this issue?  I have it set to 1h cache refresh but this confuses people who expect it to be instant.  I don't want to make it 1m since it appears to pull all charts on refresh which might cost us in storage egress fees.

**Comment by Andrea-Gallicchio** [comment]
@jdolitsky which release will contain the fix? v0.16?
